intra/android/tun2socks.go

```diff
@@ -15,59 +15,65 @@
 package tun2socks
 
 import (
-	"runtime/debug"
+	"errors"
+	"io"
+	"io/fs"
+	"log"
+	"os"
 	"strings"
 
 	"github.com/Jigsaw-Code/outline-go-tun2socks/intra"
 	"github.com/Jigsaw-Code/outline-go-tun2socks/intra/doh"
 	"github.com/Jigsaw-Code/outline-go-tun2socks/intra/protect"
-	"github.com/Jigsaw-Code/outline-go-tun2socks/tunnel"
-	"github.com/eycorsican/go-tun2socks/common/log"
+	"github.com/Jigsaw-Code/outline-sdk/network"
 )
 
-func init() {
-	// Conserve memory by increasing garbage collection frequency.
-	debug.SetGCPercent(10)
-	log.SetLevel(log.WARN)
-}
-
 // ConnectIntraTunnel reads packets from a TUN device and applies the Intra routing
 // rules.  Currently, this only consists of redirecting DNS packets to a specified
 // server; all other data flows directly to its destination.
 //
 // `fd` is the TUN device.  The IntraTunnel acquires an additional reference to it, which
-//     is released by IntraTunnel.Disconnect(), so the caller must close `fd` _and_ call
-//     Disconnect() in order to close the TUN device.
+//
+//	is released by IntraTunnel.Disconnect(), so the caller must close `fd` _and_ call
+//	Disconnect() in order to close the TUN device.
+//
 // `fakedns` is the DNS server that the system believes it is using, in "host:port" style.
-//   The port is normally 53.
+//
+//	The port is normally 53.
+//
 // `udpdns` and `tcpdns` are the location of the actual DNS server being used.  For DNS
-//   tunneling in Intra, these are typically high-numbered ports on localhost.
+//
+//	tunneling in Intra, these are typically high-numbered ports on localhost.
+//
 // `dohdns` is the initial DoH transport.  It must not be `nil`.
 // `protector` is a wrapper for Android's VpnService.protect() method.
 // `listener` will be provided with a summary of each TCP and UDP socket when it is closed.
 //
 // Throws an exception if the TUN file descriptor cannot be opened, or if the tunnel fails to
 // connect.
-func ConnectIntraTunnel(fd int, fakedns string, dohdns doh.Transport, protector protect.Protector, listener intra.Listener) (intra.Tunnel, error) {
-	tun, err := tunnel.MakeTunFile(fd)
+func ConnectIntraTunnel(fd int, fakedns string, dohdns doh.Transport, protector protect.Protector, listener intra.Listener) (*intra.Tunnel, error) {
+	tun, err := makeTunFile(fd)
 	if err != nil {
 		return nil, err
 	}
-	dialer := protect.MakeDialer(protector)
-	config := protect.MakeListenConfig(protector)
-	t, err := intra.NewTunnel(fakedns, dohdns, tun, dialer, config, listener)
+	t, err := intra.NewTunnel(fakedns, dohdns, tun, protector, listener)
 	if err != nil {
 		return nil, err
 	}
-	go tunnel.ProcessInputPackets(t, tun)
+	go copyUntilEOF(t, tun)
+	go copyUntilEOF(tun, t)
```

Comment:
Close the tun device once the IPDevice closes. The IPDevice shouldn't know about the tun device. See comments on tunnel.go
```suggestion
	go func() {
	    copyUntilEOF(tun, t)
	    tun.Close()
	}()
```

Reply:
Let's postpone this to next PR(s) in Intra repository, see the refactoring comment above.